vmp/resources/waf/rate_rule.go

```diff
@@ -0,0 +1,345 @@
+
+	"github.com/hashicorp/terraform-plugin-sdk/v2/diag"
+	"github.com/hashicorp/terraform-plugin-sdk/v2/helper/schema"
+	"github.com/hashicorp/terraform-plugin-sdk/v2/helper/validation"
+
+	sdkwaf "github.com/EdgeCast/ec-sdk-go/edgecast/waf"
+)
+
+func ResourceRateRule() *schema.Resource {
+
+	return &schema.Resource{
+		CreateContext: ResourceRateRuleCreate,
+		ReadContext:   ResourceRateRuleRead,
+		UpdateContext: ResourceRateRuleUpdate,
+		DeleteContext: ResourceRateRuleDelete,
+
+		Schema: map[string]*schema.Schema{
+			"account_number": {
+				Type:        schema.TypeString,
+				Optional:    true,
+				Description: "Identifies your account by its customer account number.",
+			},
+			"duration_sec": {
+				Type:         schema.TypeInt,
+				Required:     true,
+				ValidateFunc: validation.IntAtLeast(1),
+				Description: "Indicates the length, in seconds, of the rolling window that tracks the number of requests eligible for rate limiting. \\\n" +
+					"The rate limit formula is calculated through the num and duration_sec properties as indicated below. \\\n" +
+					"    `num` requests per `duration_sec` \\\n    Valid values are: \\\n    `1 | 5 | 10 | 30 | 60 | 120 | 300`",
+			},
+			"disabled": {
+				Type:     schema.TypeBool,
+				Optional: true,
+				Description: "Indicates whether this rate rule will be enforced. \\\n" +
+					"Valid values are: \n" +
+					"    * true: Disabled. This rate limit will not be applied to traffic.\n" +
+					"    * false: Enabled. Traffic is restricted to this rate limit.",
+			},
+			"name": {
+				Type:         schema.TypeString,
+				Optional:     true,
+				ValidateFunc: validation.StringIsNotWhiteSpace,
+				Description:  "Assigns a name to this access rule.",
+			},
+			"num": {
+				Type:         schema.TypeInt,
+				Required:     true,
+				ValidateFunc: validation.IntAtLeast(1),
+				Description: "Indicates the rate limit value. This value identifies the number of requests that will trigger rate limiting. \\\n" +
+					"The rate limit formula is calculated through the num and duration_sec properties as indicated below. \\\n" +
+					"`num` requests per `duration_sec`",
+			},
+			"keys": {
+				Type:     schema.TypeList,
+				Optional: true,
+				Description: "Indicates the method by requests will be grouped for the purposes of this rate rule. \\\n" +
+					"Valid values are: \n" +
+					"    * Missing / Empty Array: If the `keys` property is not defined or set to an empty array, " +
+					"all requests will be treated as a single group for the purpose of rate limiting. \n" +
+					"    * IP: Indicates that requests will be grouped by IP address. Each unique IP address is considered a separate group. \n" +
+					"    * USER_AGENT: Indicates that requests will be grouped by a client's user agent. " +
+					"Each unique combination of IP address and user agent is considered a separate group. \n",
+				Elem: &schema.Schema{
+					Type: schema.TypeString,
+				},
+			},
+			"condition_group": {
```

Comment:
I am not 100% sure but if "id" is optional, you may encounter a weird update behavior.
It would be better to create a conditional_group resource and add a reference to it.

Reply:
That's an interesting point. I will ask Jairo about the behavior of that ID property. It seems to be just metadata but I will confirm with him.